crates/component-macro/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+                                if "name" == &key.to_string() && '=' == op.as_char() =>
+                            {
+                                if name.is_none() {
+                                    name = Some(TokenTree::Literal(value.clone()));
+                                } else {
+                                    return Err(Error::new(
+                                        ident.span(),
+                                        "duplicate field rename attribute",
+                                    ));
+                                }
+                            }
+                            _ => {
+                                return Err(Error::new_spanned(
+                                    &attribute.tokens,
+                                    UNEXPECTED_SYNTAX_MESSAGE,
+                                ))
+                            }
+                        }
+                    }
+                    _ => {
+                        return Err(Error::new_spanned(
+                            &attribute.tokens,
+                            UNEXPECTED_SYNTAX_MESSAGE,
+                        ))
+                    }
+                }
+            }
+        }
+    }
+
+    Ok(name)
+}
+
+fn add_trait_bounds(mut generics: syn::Generics) -> syn::Generics {
+    for param in &mut generics.params {
+        if let syn::GenericParam::Type(ref mut type_param) = *param {
+            type_param
+                .bounds
+                .push(parse_quote!(wasmtime::component::ComponentType));
+        }
+    }
+    generics
+}
+
+#[proc_macro_derive(Lift, attributes(component))]
+pub fn lift(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    expand_lift(parse_macro_input!(input as DeriveInput))
+        .unwrap_or_else(Error::into_compile_error)
+        .into()
+}
+
+fn expand_lift(input: DeriveInput) -> Result<TokenStream> {
+    match find_style(&input)? {
+        Style::Record => expand_record_lift(input),
+
+        style => Err(Error::new_spanned(
+            input,
+            format!("todo: expand_lift {style:?}"),
+        )),
+    }
+}
+
+fn expand_record_lift(input: DeriveInput) -> Result<TokenStream> {
+    let name = &input.ident;
+
+    let body = if let Data::Struct(body) = &input.data {
+        body
+    } else {
+        return Err(Error::new(name.span(), INCOMPATIBLE_TYPE_MESSAGE));
+    };
+
+    match &body.fields {
+        syn::Fields::Named(fields) => {
+            let mut types = TokenStream::new();
+            let mut field_initializations = TokenStream::new();
+
+            for (index, syn::Field { ty, ident, .. }) in fields.named.iter().enumerate() {
+                types.extend(quote!(#ty,));
+
+                let index = syn::Index::from(index);
+
+                field_initializations.extend(quote!(#ident: result.#index,));
+            }
+
+            delegate_record_lift(
+                input,
+                quote!((#types)),
```

Comment:
Oh this is clever! I hadn't actually thought of simply delegating to the tuple implementation. That definitely makes this easier since you don't have to do offset calculations here in the derive macro.

That being said though the main limitation of this is that it won't work for structs with > 16 fields. That doesn't seem too common though so otherwise seems like a good way to start?

Reply:
One thing you may want to explore though is what the error messages look like if a field of the record doesn't implement the requisite trait. For example what's the compiler error message when you derive `Lift` for a record with a `String` in it?

Reply:
The cleverness is due to @jameysharp -- that's straight from [his patch](https://github.com/bytecodealliance/wasmtime/compare/main...jameysharp:component-derive).

Reply:
I'm certainly enjoying being called clever! But before I got sidetracked, I had intended to drop this trick. :laughing:

In #4217 I factored out `next_field` to make both tuple and record macro expansions relatively straightforward. While `lower` and `lift` were already pretty easy, `next_field` wraps up the common math for `size`, `store`, and `load`. So I hope it should be simple now to almost just copy `impl_component_ty_for_tuples`, except as a proc-macro instead of `macro_rules`, and with field names instead of indexes.

As I recall, this bit of cleverness is also what forced me to call `.clone()` in `lower`: I couldn't build a tuple of borrows, because the implementation I delegated to expected the borrow to be on the outside of the tuple instead. Skipping the indirection should let you remove those clones.